### PR TITLE
Auto updates

### PR DESCRIPTION
**Pull request by jinroh**
This PR adds the auto updates process. The instance struct has a new `AutoUpdate` field to activate the auto update.

The admin API has new entries to execute manually the auto-updates. The configuration also has a new block to configure the auto-updates scheduler.

~~Still missing: /settings route to set an instance `AutoUpdate`. We also coud decide to make it the default value.~~

### CODE REVIEW

File: web/jsonapi/data.go [hunk 1]
```diff
@@ -94,13 +94,6 @@ func MarshalObject(o Object) (json.RawMessage, error) {
 	links := o.Links()
 	rels := o.Relationships()
 
-	o.SetID("")
-	o.SetRev("")
-	defer func() {
-		o.SetID(id)
-		o.SetRev(rev)
-	}()
-
```
Comment (jinroh) [inline web/jsonapi/data.go]: @nono

I removed this hack because it provoked some race-conditions that would break installation APIs that rely on the fact that the manifest should not be mutated by consumer of the API. Their may be other race-conditions of the same nature in the code.

The side effect is that the `_id` and `_rev` in the `data` field.

File: cmd/instances.go [hunk 2]
```diff
@@ -437,6 +437,36 @@ var oauthClientInstanceCmd = &cobra.Command{
 	},
 }
 
+var updateAllCmd = &cobra.Command{
+	Use:   "update-all [slugs...]",
```
Comment (nono) [inline cmd/instances.go]: Maybe we can use an `--all-domains` flag like many other commands

File: cmd/instances.go [hunk 3]
```diff
@@ -437,6 +437,36 @@ var oauthClientInstanceCmd = &cobra.Command{
...
+	RunE: func(cmd *cobra.Command, args []string) error {
+		c := newAdminClient()
+		return c.Updates(&client.UpdatesOptions{
+			Slugs: args,
+		})
+	},
+}
+
+var updateCmd = &cobra.Command{
+	Use:   "update [domain] [slugs...]",
```
Comment (nono) [inline cmd/instances.go]: What about the `cozy-stack apps update` command? Does this command replace it? Or are they are complementary?

File: pkg/apps/webapp.go [hunk 4]
```diff
@@ -398,7 +401,7 @@ func GetWebappBySlug(db couchdb.Database, slug string) (*WebappManifest, error)
 // TODO: pagination
 func ListWebapps(db couchdb.Database) ([]*WebappManifest, error) {
 	var docs []*WebappManifest
-	req := &couchdb.AllDocsRequest{Limit: 100}
+	req := &couchdb.AllDocsRequest{Limit: 200}
```
Comment (nono) [inline pkg/apps/webapp.go]: Why this change? Are we near the limit of 100?

File: pkg/instance/instance.go [hunk 5]
```diff
@@ -76,11 +76,12 @@ var (
...
-	DocRev string `json:"_rev,omitempty"` // couchdb _rev
-	Domain string `json:"domain"`         // The main DNS domain, like example.cozycloud.cc
-	Locale string `json:"locale"`         // The locale used on the server
-	Dev    bool   `json:"dev"`            // Whether or not the instance is for development
+	DocID        string `json:"_id,omitempty"`  // couchdb _id
+	DocRev       string `json:"_rev,omitempty"` // couchdb _rev
+	Domain       string `json:"domain"`         // The main DNS domain, like example.cozycloud.cc
+	Locale       string `json:"locale"`         // The locale used on the server
+	NoAutoUpdate bool   `json:"no_auto_update"` // Whether or not the instance has auto updates for its applications
+	Dev          bool   `json:"dev"`            // Whether or not the instance is for development
```
Comment (nono) [inline pkg/instance/instance.go]: Maybe we should use `omitempty` for `NoAutoUpdate` and `Dev`

File: pkg/instance/updates.go [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+	}
+
+	u := &updateCron{
+		stopped:  make(chan struct{}),
+		finished: make(chan struct{}),
+	}
+
+	scheduleValue := autoUpdates.Schedule
+	if scheduleValue == "" {
+		scheduleValue = "@midnight"
```
Comment (nono) [inline pkg/instance/updates.go]: I thought `""` means that the auto update is disabled, https://github.com/cozy/cozy-stack/pull/948/files#diff-091ace02b49a46f55f2586cfec89a160R306

File: cozy.example.yaml [hunk 7]
```diff
@@ -76,6 +76,10 @@ mail:
 # directory with the hooks scripts - flags: --hooks
 hooks: ./scripts/hooks
 
+# Auto updates scheduler
+auto_updates:
+  schedule: "@cron 0 0 0 * * *"
```
Comment (nono) [inline cozy.example.yaml]: For a cluster, only one node should have the `auto_updates.schedule`? Or should all node have it?

File: pkg/instance/updates.go [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+// UpdateInstance starts the auto-update process on the given instance. The
+// slugs parameters can be used to filter (whitelist) the applications' slug
+func UpdateInstance(inst *Instance, slugs ...string) error {
+	insc := make(chan *apps.Installer)
+	errc := make(chan error)
+
+	var g sync.WaitGroup
+	g.Add(4)
```
Comment (nono) [inline pkg/instance/updates.go]: What is this `4`?

File: web/jsonapi/data.go (hunk 1 above)
Comment (nono) [reply web/jsonapi/data.go]: I don't like that `_id` and `_rev` are present in the `data.attributes` field.

File: web/settings/instance.go [hunk 9]
```diff
@@ -68,9 +69,20 @@ func updateInstance(c echo.Context) error {
 		return err
 	}
 
+	var needUpdate bool
```
Comment (nono) [inline web/settings/instance.go]: [nitpick] I'd prefer `needUpdate := false`

File: cmd/instances.go [hunk 10]
```diff
@@ -437,6 +437,36 @@ var oauthClientInstanceCmd = &cobra.Command{
...
+	Short: "Starts the updates for the specified domain instance.",
+	Long: `Starts the updates for the specified domain instance. The slugs
```
Comment (nono) [inline cmd/instances.go]: s/Starts/Start/ to be coherent with other commands

File: cozy.example.yaml (hunk 7 above)
Comment (jinroh) [reply cozy.example.yaml]: For a cluster, I think it is better to control the auto updates via the administration API to have better supervision of when starting the autoupdates.
